/*
 * CS-6200 Preparation - Module 03: gRPC Service Definition
 *
 * Defines a simple key-value store service.
 * This is similar to what you'll build in GIOS Project 4.
 *
 * Generate code:
 *   protoc --grpc_out=. --cpp_out=. --plugin=protoc-gen-grpc=`which grpc_cpp_plugin` keyvalue.proto
 */

syntax = "proto3";

package keyvalue;

// The key-value store service
service KeyValueStore {
    // Unary RPC: Get a value by key
    rpc Get(GetRequest) returns (GetResponse);

    // Unary RPC: Put a key-value pair
    rpc Put(PutRequest) returns (PutResponse);

    // Unary RPC: Delete a key
    rpc Delete(DeleteRequest) returns (DeleteResponse);

    // Server streaming: Get all keys
    rpc ListKeys(ListKeysRequest) returns (stream KeyResponse);

    // Client streaming: Put multiple key-value pairs
    rpc PutMultiple(stream PutRequest) returns (PutMultipleResponse);
}

// Request/Response messages

message GetRequest {
    string key = 1;
}

message GetResponse {
    bool found = 1;
    string value = 2;
}

message PutRequest {
    string key = 1;
    string value = 2;
}

message PutResponse {
    bool success = 1;
    bool was_update = 2;  // true if key existed
}

message DeleteRequest {
    string key = 1;
}

message DeleteResponse {
    bool success = 1;
    bool existed = 2;  // true if key existed before delete
}

message ListKeysRequest {
    string prefix = 1;  // Optional: filter keys by prefix
}

message KeyResponse {
    string key = 1;
}

message PutMultipleResponse {
    int32 count = 1;  // Number of keys inserted
}
